include/xlnt/utils/datetime.hpp

```diff
@@ -131,6 +135,11 @@ struct XLNT_API datetime
     /// The microsecond
     /// </summary>
     int microsecond;
+
+    /// <summary>
+    /// Whether the date is in an empty state.
+    /// </summary>
+    bool is_null = true;
```

Comment:
Using a boolean for null state just increases class size unnecessarily. I think using a magic 0/0/0 as day/month/year is fine, as long as this magic number is internal to the class.

Reply:
To be precise, it increases `xlnt::date` by 4 bytes (1 byte for the boolean and 3 bytes for padding) to a total of 16 bytes. Also, it increases `xlnt::datetime` by 4 bytes (1 boolean, 3 padding) to a total of 32 bytes.

However, I personally disagree that it is not necessary. Having a magic number for that purpose - even if it is internal - means that a user of `xlnt::date` or `xlnt::datetime` would never be allowed to use the magic date, as it would mean that the value is invalid. Even worse, since this is internal, the user would never have the possibility to check it either, but would be required to read the internals of our classes.

While writing correct code would be possible without an additional variable, I think our `xlnt::date` and `xlnt::datetime` have not really been made for that (not yet, at least). Examples from C++20:

- [`std::chrono::day` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/day/ok) function that checks whether a day is in the allowed range
- [`std::chrono::month` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/month/ok) function that checks whether a month is in the allowed range
- [`std::chrono::year` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/year/ok) function that checks whether a year is in the allowed range
- [`std::chrono::year_month_day` has an `ok()`](https://en.cppreference.com/w/cpp/chrono/year_month_day/ok) that checks whether the year, month and day are in range (like above), and further checks things like leap years (whether February 29 is valid in the specified year) and whether 31st of a month is allowed in the specified month.

The new C++20 structures allow this **_without an additional variable_** by specifying very precise rules on how to handle out of range values. However, we're only using `int`s for all 3 fields, and we're not enforcing any ranges or checking them at all. In other words, we currently have no validation, so we also cannot do any sanity checks. A magic number would make the situation even worse.

Besides, if we really wanted to decrease the class size we could use `unsigned char` / `uint8_t` for the month, day, hour, minute, second. With an adapted order of the fields (higher-sized types at the top - that is, year and microsecond) that would result in a class size of 16 bytes for `xlnt::datetime` and a size of 8 for `xlnt::date`.

Reply:
Ok, let's keep it like this for now.

Reply:
Another thing that I do like even less is that we expose this `is_null` member publicly. (sorry that I didn't catch it on first review). I know the other members are also public (and I think we should change this in the next major release), but storing is_null as a boolean really seems to me as an implementation detail.
What do you think about changing this to a private member and provide getter/setter functions?
Another inconvenience is that after setting year/month/day the user has also to explicitly set the `is_null` to false. Maybe we should already add setter/getter functions for year/month/day and automatically set is_null to false (and already deprecate access to the public member variables).

Reply:
I agree, and I dislike it as well. My newest commit has changed that and has also deprecated directly accessing the member variables (for now only as a code comment).

I agree to your other remarks. Since I'm a big fan of simple solutions whenever they are possible (as long as they are correct, obviously), I would propose the following solution that I implemented as part of my newest commit:

1. `xlnt::date` and `xlnt::datetime` are no longer meant for direct access to their member variables (which is still allowed for now, but using the getters is encouraged). From now on, they are no longer meant to be modified on the fly: changing an `xlnt::date` or `xlnt::datetime` should only be done by constructing a new one, which is very cheap performance-wise. This whole idea makes a lot of things much easier, as we only have to handle the initial state correctly, but state can no longer be changed on the fly. This also avoids us having to provide setters. It is similar to [`std::chrono::year_month_day`](https://en.cppreference.com/w/cpp/chrono/year_month_day), since it also only has getters but no setters.
2. Both `xlnt::date` and `xlnt::datetime` can now be default-constructed, in which case they are initialized with an empty/invalid state.
3. Both `xlnt::date` and `xlnt::datetime` can be initialized using integers for the year/month/day etc. - however, like before, we still perform no checks on the members, but assume that users know what they are doing and simply assume that the state is NOT empty/invalid.

So, in short, we can either have default-constructed instances having empty/invalid states, or programmer-constructed instances always having valid states. Also, the whole state can no longer change after construction (or should not beginning with the next major release). This makes the code much simpler.

Reply:
I'm fine with a const date(time) object.